cmd/podman/completion/completion.go

```diff
@@ -0,0 +1,368 @@
+package completion
+
+import (
+	"context"
+	"strings"
+
+	"github.com/containers/libpod/cmd/podman/common"
+	"github.com/containers/libpod/cmd/podman/registry"
+	"github.com/containers/libpod/cmd/podman/validate"
+	"github.com/containers/libpod/pkg/domain/entities"
+	"github.com/containers/libpod/pkg/registries"
+	systemd "github.com/containers/libpod/pkg/systemd/generate"
+	"github.com/spf13/cobra"
+)
+
+var (
+	completionOptions = entities.CompletionOptions{}
+	completionCmd     = &cobra.Command{
+		Use:              "completion",
+		Short:            "Generate shell autocompletions",
+		Args:             cobra.ExactArgs(1),
+		Long:             "Generate shell autocompletions for bash/zsh/fish/powershell\nThis uses the internal cobra functions",
+		TraverseChildren: true,
+		RunE:             validate.SubCommandExists,
+		//dont show command to user
+		Hidden: true,
+	}
+)
+
+func init() {
+	registry.Commands = append(registry.Commands, registry.CliCommand{
+		Mode:    []entities.EngineMode{entities.ABIMode, entities.TunnelMode},
+		Command: completionCmd,
+	})
+}
+
+//Get containers with status.
+func getContainers(status string) ([]string, cobra.ShellCompDirective) {
+	var suggestions []string
+
+	listOpts := entities.ContainerListOptions{
+		Filters: make(map[string][]string),
+	}
+	listOpts.All = true
+
+	if status != "all" {
+		listOpts.Filters = map[string][]string{"status": {status}}
+	}
+
+	containers, err := registry.ContainerEngine().ContainerList(registry.GetContext(), listOpts)
+	if err != nil {
+		return nil, cobra.ShellCompDirectiveError
+	}
+
+	for _, container := range containers {
+		//include name in suggestions
+		//if ids should be also completed
+		//suggestions = append(suggestions, container.Names[0], container.ID[0:12])
+		suggestions = append(suggestions, container.Names...)
+	}
+
+	return suggestions, cobra.ShellCompDirectiveNoFileComp
+}
+
+//Get pods with status.
+func getPods(status string) ([]string, cobra.ShellCompDirective) {
+	var suggestions []string
+
+	listOpts := entities.PodPSOptions{
+		Filters: make(map[string][]string),
+	}
+
+	if status != "all" {
+		listOpts.Filters = map[string][]string{"status": {status}}
+	}
+
+	pods, err := registry.ContainerEngine().PodPs(registry.GetContext(), listOpts)
+	if err != nil {
+		return nil, cobra.ShellCompDirectiveError
+	}
+
+	for _, pod := range pods {
+		//include name in suggestions
+		//if ids should be also completed
```

Comment:
Yes. Container IDs aren't usually user-visible, but I can attest that it is reeeeally convenient to `podman ps`, see a container I want to operate on, and type just the first two characters of the SHA and want completion on it.